### PR TITLE
Added support for Windows. Tested on Windows 7 64bit

### PR DESCRIPTION
**Pull request by yerpj**
I modified the initial script in order to make it compatible with Windows. If on Windows, it uses pipes instead of fifo. 
Test setup:
1x 802.15.4 node
1x 802.15.4 router
1x sniffer on Serial Port
Wireshark 2.2

### CODE REVIEW

**Comment by g-oikonomou** [comment]
This is a very desirable feature, but not sure about some of the details. See inline comments please.

Also, I indent with 4 spaces, looks like you've used tabs here? We'd have to change this.

File: sensniff.py [hunk 1]
```diff
@@ -45,23 +45,43 @@
...
+	defaults = {
+		'device': '/dev/ttyUSB0',
+		'baud_rate': 460800,
+		'out_file': 'sensniff.hexdump',
+		'out_fifo': '/tmp/sensniff',
+		'out_pcap': 'sensniff.pcap',
+		'debug_level': 'WARN',
+		'log_level': 'INFO',
+		'log_file': 'sensniff.log',
+	}
```
Comment (g-oikonomou) [inline sensniff.py]: Looks like the only change here is for 'device', so can we just use a separate variable and then keep a single defaults map? Would make the code more readable I think.

File: sensniff.py [hunk 2]
```diff
@@ -330,22 +368,40 @@ def handle(self, data):
...
+					logger.info('Wrote a frame of size %d bytes' % (data.len))
+					stats['Piped'] += 1
+				except IOError as e:
+					if e.errno == errno.EPIPE:
+						logger.info('Remote end stopped reading')
+						stats['Not Piped'] += 1
+						self.of = None
+						self.needs_pcap_hdr = True
+					else:
+						raise
```
Comment (g-oikonomou) [inline sensniff.py]: Again, not hugely keen at those two mega if/else blocks. Looks like the exception part is ideantical across the two. I think what we need is inherit `FifoOutHandler` for linux/win and provide correct `__init__`, FIFO/Pipe creation and `handle`

Comment (yerpj) [inline sensniff.py]: I agree. Actually, this first PR aims at providing a basic 'just working' script on Windows. I am totally beginner with Python. Maybe someone more comfortable with Python may re-implement this section in order to make things look beautiful. What I propose is to close this PR if you expect a better implementation, and I will keep things 'as is' on my fork.

File: sensniff.py (hunk 1 above)
Comment (yerpj) [inline sensniff.py]: mostly done

**Comment by yerpj** [comment]
done for the "TABS instead of spaces" as well as for the 'default' structure

**Comment by g-oikonomou** [comment]
Many thanks! I'll test things and see how to improve code structure.

**Comment by yerpj** [comment]
BTW, I remember that I encountered sometimes an issue with both the current version (1.2)as well as with my modified script (1.2.1) . The issue, raised from `logger.info("Peripheral: %s%s" % (b.decode(), per_out.decode()))`, said that an input character , which was not 'ascii' (as far as I know, my sniffer sends ascii only bytes), could not  be parsed, or something more or less similar.

**Comment by g-oikonomou** [comment]
Hmmm I think I know what you are talking about. Please keep an eye out and fire away an issue if you encounter this again.

Can I assume you are running Contiki on the device? May I ask what h/w?

**Comment by yerpj** [comment]
Actually I don't run Contiki. We run our own system on our own nodes. I implemented a 'packet-injector' task thanks to your well documented [README](https://github.com/g-oikonomou/sensniff#sensniff-host-to-peripheral-protocol), that basically sends every received packet to sensniff. This task does not interract at all, only sends every packet.

**Comment by g-oikonomou** [comment]
This is great, this is the first time I've heard someone using this with custom firmware! :)

**Comment by g-oikonomou** [comment]
Joking aside, this will be tricky to handle. The script reads in 5-byte chunks, expecting to see the magic + version bytes. When this condition occurs, those reads will only become aligned to the start of a frame when the peripheral has stayed silent for more than the timeout. This may end up taking quite a long time, with many frames discarded in between.

Gets even worse. Once the script reads the correct header, it will then try to read out the entire remainder of the frame in one go (near line 226 `b = self.port.read(length)`). This has to happen within the timeout. For the current timeout of 100ms at the default of 460800, the theoretical maximum is 46080 bits =  5760 bytes. In the future I want to be able to support long 802.15.4g frames (up to 2KB), so this condition may become more frequent.

However, a good start is to get the script to handle this condition without blowing up and I have pushed branch `fix-unicode-error`. My testbed is not set up at the moment, so could I ask you to revert timeout to 0.1 and try with this branch?

**Comment by g-oikonomou** [comment]
When creating the pipe on win, are you actually getting a file created on disk?

**Comment by yerpj** [comment]
The pipe is not seen on the file system, unfortunately. It is created at specific path `\\.\pipe` like every pipes.
You can verify that the pipe `wireshark`has been created by using [this tool](https://technet.microsoft.com/en-us/sysinternals/dd581625.aspx)

Warning, the pipe has to be created and feeded with pcap header at first, before wireshark times out. If wireshark doesn't get the pcap header after a few milliseconds, it returns an error of the kind `can't find the specified pipe`

The reference implementation for wireshark can be consulted [HERE](https://wiki.wireshark.org/CaptureSetup/Pipes)

**Comment by g-oikonomou** [comment]
Yep, I was just reading that wiki page too.

So you are telling me there is basically no way of opening the pipe by adding it in the respective wireshark dialogue, right?

**Comment by yerpj** [comment]
Actually you can add a pipe under the respective dialog, but as far as I remember, wireshark will not be able to capture packets on it because it will nor recognize this pipe as a pcap feeder....To be confirmed.

I thought it was convenient for the user to only run the script, and let it open and configure wireshark accordingly.

**Comment by g-oikonomou** [comment]
> I thought it was convenient for the user to only run the script, and let it open and configure wireshark accordingly.

On one hand yes, on the other hand it is nice to be able to stop and restart WS while the script is still running. This can be used to e.g. clean very large capture sessions: I've had sessions with 100s of Ks of captured frames and it's nice to reduce the amount of scrolling around.

On unix this is pretty easy to achieve, since when wireshark stops the script can detect that nothing is reading the fifo and it will then write the global pcap header when wireshark starts again. Not sure if this behaviour can be achieved with windows named pipes though.

**Comment by yerpj** [comment]
I would be very surprised if Windows could achieve such a behavior ;-)

Actually I don't know at all. I agree with you, the current behavior on Unix platform can be very convenient,  I hope someone skilled in Windows platforms could help us to achieve the same functionalities

**Comment by g-oikonomou** [comment]
I have have added branch `win-support`. I've kept your commits from #11, plus:

* Code structure cleanup to avoid those multiple platform tests.
* Support for the `W` / `--ws-path` argument (only appears on Windows), so the user can specify an alternative wireshark installation path
* Deserves a minor version increase :)
* Some documentation

In line 74, we may need to prefix the string with 'r'. Please would be kind enough to test this thing actually works and also to try out `-W`. I am not 100% certain whether the path will need enclosed in quotes or whatnot.

**Comment by yerpj** [comment]
I tested on branch `win-support`. The COM is correctly opened. The pipe is created, too, but I suspect that the pcap header is not sent over the pipe, Wireshark does not seem to be able to connect on it.

**Comment by g-oikonomou** [comment]
Pretty certain I'm following identical logic as your original contribution: Global PCAP header written in `handle()`, therefore upon reception of first captured frame from the peripheral. Have I missed something?

**Comment by yerpj** [comment]
That may be the problem.  Why do we need to wait on first frame to send the pcap header ? In theory, WS may be running BEFORE the reception of the first frame,  right ? 
You are right,  after the first packet,  WS can capture on the pipe.

**Comment by g-oikonomou** [comment]
We don't need to wait, no. Since WS will be open on Win, we can basically write the header immediately after connecting to the pipe.

On Linux / OS X, the user can start/stop, open/close WS at will. This is why we write the global pcap header in `handle()`

**Comment by yerpj** [comment]
if I resume, steps are :
1) run script sensniff.py--> it will create the pipe
2) at first incoming packet, the script sends the pcap header into pipe
3) WS can now capture on the pipe

I expected to call `win32file.WriteFile(self.of, pcap_global_hdr)` right after the pipe creation within the __init__ method, thus allowing wireshark to be configured to listen on the pipe even if no packet is already received.

**Comment by g-oikonomou** [comment]
On Linux / OS X

1. Create FIFO, connect to FIFO
2. Set flag that we need to write the global PCAP header before next frame
2. Capture frame, check if the FIFO has a remote end reading. 
2a. If not, discard frame and go to 2.
2b. If so, check whether we need to write the global header. If so, write the header. Set flag that we no longer need the header. Write the frame to the FIFO.

The user can request a new global PCAP header at any time in interactive mode.

On Win: What you are describing is what the logic currently does. Depending on how long WS will wait for the global PCAP header, we may have to slightly modify the logic to something like:

1. Create Pipe, connect to pipe
2. Run WS
3. Write global header
4. Wait for frame, write frame to pipe

As you say above, yes, steps 1-3 would have to be inside `__init__`

**Comment by yerpj** [comment]
I just ran the script using my laptop (branche: `win-support`, os: Windows 7sp1, architecture: 64-bit) and the script seems to hang. 
command: `python sensniff.py -d COM9 -b 9600 -D INFO`
logs:
```
Started logging
Serial port COM9 opened
Named pipe correctly created
```
I verified, the COM port is correctly opened, and the pipe is created, too. But the script does not forward any packets and the `CTRL+V` command doesn't kill the script. I have to close the cmd.exe window at each time. 
Do you have any idea on what may be happening?

**Comment by g-oikonomou** [comment]
One immediate comment is that 9600 is way too low.

Did you mean to type Ctrl+C?

**Comment by yerpj** [comment]
9600 was fine with branche `master` 

Yes, I meant `CTRL+C`

**Comment by g-oikonomou** [comment]
I strongly advise getting your hardware's driver to support 460800 at least, otherwise the serial interface will become a bottleneck and you will start missing packets / getting corrupt packets.

I cannot immediately think of any other reasons for the script to crash. As you will be able to see, the logic is identical to your original pull.

**Comment by yerpj** [comment]
With the real sniffer that I am using those days, serial port is at 115200 baud, and the incoming packets are being buffered into the sniffer if needed.
Right now, I am emulating a sniffer with a packet injector, and the HW used (Espruino Pico) is limited to 9600 Baud on serial output. Anyway, I used to use the exact same setup a few days ago in order to figure out that this [issue](https://github.com/g-oikonomou/sensniff/issues/12) was triggered by a timeout on the host side (python serial timeout). The baudrate was not the problem, really. I will retry on my desktop tomorrow to find out what's wrong with this particular setup.

**Comment by g-oikonomou** [comment]
OK yeah gotcha, the baud rate will only become a problem if you are receiving wireless traffic faster than you can stream it down the serial line.

Perhaps running with `-D INFO` will help

**Comment by yerpj** [comment]
> Perhaps running with `-D INFO` will help

This was the case on last try

**Comment by yerpj** [comment]
I just had another test on my desktop this time. Same behavior. 
But I tried something else:
1. Run WS (manually), add pipe `\\.\pipe\wireshark`
2. Run script sensniff.py 
3. Ask WS to begin capture on the pipe 

This way, wireshark, the solution seems to work. At least the sniffed packets are displayed on the capture window of WS. 
I wonder why WS is not opened from the script itself (`proc=subprocess.Popen([os.path.join(ws_path,'Wireshark.exe'),'-i' + self.named_pipe, '-k'])`)

**Comment by g-oikonomou** [comment]
So WS not launching properly before the hang? Can you debug the value of this? `os.path.join(ws_path,'Wireshark.exe')`

**Comment by yerpj** [comment]
Actually it seems that it never executes this instruction. The script seems to hang while trying to execute `win32pipe.ConnectNamedPipe(pipe, None)`.

**Comment by g-oikonomou** [comment]
!

**Comment by yerpj** [comment]
Sorry, I could have explained a little more:
here is how I tried to understand what was hanging:
```
        try:
            pipe = win32pipe.CreateNamedPipe(self.named_pipe,
                                             win32pipe.PIPE_ACCESS_OUTBOUND,
                                             win32pipe.PIPE_TYPE_MESSAGE |
                                                 win32pipe.PIPE_WAIT,
                                             1, 65536, 65536, 300, None)
            logger.info('Named pipe correctly created')
            logger.info('HERE')
            win32pipe.ConnectNamedPipe(pipe, None)
            logger.info('THERE')
            self.of = pipe
            proc = subprocess.Popen([os.path.join(ws_path, 'Wireshark.exe'),'-i' + self.named_pipe, '-k'])			
```

```
sensniff>python sensniff.py -d COM43 -b 9600 -D INFO
Started logging
Serial port COM43 opened
Named pipe correctly created
HERE
```

As I said, the pipe is correctly created, I can see it with the windows specific tool `pipelist`.
We see that the `logger.info('THERE')` is never reached.

**Comment by g-oikonomou** [comment]
Yep, I understand. I'm just surprised it doesn't connect since the code is basically a raw copy-paste from your original pull. Can't think of what I may have missed

**Comment by yerpj** [comment]
I think I got it! WS has to be opened BEFORE connecting to pipe. Maybe the pipe has to already have a 'consumer' in order to let `win32pipe.ConnectNamedPipe(pipe,None)` execute as expected. 
Anyway, by modifying the __init__ method this way:
```
            proc = subprocess.Popen([os.path.join(ws_path, 'Wireshark.exe'),'-i' + self.named_pipe, '-k'])
            pipe = win32pipe.CreateNamedPipe(self.named_pipe,
                                             win32pipe.PIPE_ACCESS_OUTBOUND,
                                             win32pipe.PIPE_TYPE_MESSAGE |
                                                 win32pipe.PIPE_WAIT,
                                             1, 65536, 65536, 300, None)
            logger.info('Named pipe correctly created')
            win32pipe.ConnectNamedPipe(pipe, None)
            self.of = pipe				
```
instead of 
```
            pipe = win32pipe.CreateNamedPipe(self.named_pipe,
                                             win32pipe.PIPE_ACCESS_OUTBOUND,
                                             win32pipe.PIPE_TYPE_MESSAGE |
                                                 win32pipe.PIPE_WAIT,
                                             1, 65536, 65536, 300, None)
            logger.info('Named pipe correctly created')
            win32pipe.ConnectNamedPipe(pipe, None)
            self.of = pipe
            proc = subprocess.Popen([os.path.join(ws_path, 'Wireshark.exe'),'-i' + self.named_pipe, '-k'])				
```
seems to work as expected. I tested with my sniffer, and it's fine.

**Comment by g-oikonomou** [comment]
Ha!

That is one logic I _did_ change compared to your original pull. Similar behaviour to FIFOs: a process cannot write before the reading side has been opened. Shame win32pipe just gets stuck instead of throwing an exception there, we could have done fun things by handling it.

Your proposed change makes sense.

Thanks for all the input by the way, much appreciated!

**Comment by yerpj** [comment]
My pleasure !

Do you think you could apply the modifications on your side? I don't know how to handle branches on a forked project :-/

**Comment by g-oikonomou** [comment]
I can apply on my end yes

Or you can reset the branch you used to open to pull to point to my branch, then commit on top of your branch (which now has my latest changes) and push -f to your fork. This will automatically update the pull to include my version + your latest change! Let me know if you fancy fiddling, if not I'll do it when I get a moment

**Comment by yerpj** [comment]
I just made a new PR #13 that basically solves the issue discussed above. Therefore, this PR may be closed, right?

**Comment by g-oikonomou** [comment]
#13

**Comment by g-oikonomou** [comment]
Did you get a chance to look at the proposed fix in b9a746b8f280fc393b1d97e54f98fddecabac8d0?

**Comment by marcomilazzo** [comment]
hi
i get the following error:
Unexpected error: <type 'exceptions.NameError'>

I'm under windows 7  and phyton 2.7
it looks to me that it doesn' create the pipe
A problem of permissions?
i have no experiance with phyton !
Can you help me?
thank's
marco

**Pull request closed without merging**